drivers: hwinfo: Add reset cause support

arvinf opened the pull request:
This PR adds initial support for retrieving SoC reset cause through the `hwinfo` driver.

`hwinfo_get_reset_cause` and `hwinfo_clear_reset_cause` are used to retrieve and to clear the cause of system reset, on supported platforms. This feature is typically provided by the SoC using a register with bitflags corresponding to different reasons the chip was reset.
    
Different platforms can provide different causes of reset, however there is a great deal of overlap. `enum reset_cause` can be expanded in the future to support additional reasons, as additional platforms are supported.

arvinf committed:
drivers: hwinfo: Add reset cause support

Add `hwinfo_get_reset_cause` and `hwinfo_clear_reset_cause` to retrieve
and to clear cause of system reset on supported platforms.

Different platforms can provide different causes of reset, however
there is a great deal of overlap. `enum reset_cause` can be expanded in
the future to support additional reasons, as additional platforms are
supported.

Signed-off-by: Arvin Farahmand <arvinf@ip-logix.com>

alexanderwachter commented:
Hi @arvinf. Thanks for your PR! It looks very useful!
I didn‘t looked trough To it carefully yet but, from what i hav seen, the userspace handlers are missing and there are no testcases. Those two things are mandatory to get this in. In the handlers you need to check whether the thread has access to the pointer provided, or not. If you only have a return value that is negative on error and positive (flags) on success, you don‘t need any checks nor handlers. If you need help, either write here or contact me on slack.

arvinf commented:
Thank you. Where do I find information on how to add those? `hwinfo_clear_reset_cause` doesn't take any arguments and only returns an `int` result code, so it doesn't need any tests, or handlers, correct?

nashif commented:
adding @ceolin  @andrewboie  for syscalls...

tyrantlizard77 reviewed (commented):
Syscall verification functions are missing.
Documentation: https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html

Also needs tests. I don't know how much coverage you can get but we at least need to build everything and exercise these APIs.

arvinf commented:
I added verification calls, and added test cases. I kept the tests fairly simple since the API behaviour is SoC dependent.

ceolin commented on drivers/hwinfo/hwinfo_handlers.c:
```diff
@@ -14,3 +14,18 @@ ssize_t z_vrfy_hwinfo_get_device_id(u8_t *buffer, size_t length)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_WRITE(cause, sizeof(u32_t)));
+
+	return z_impl_hwinfo_get_reset_cause((u32_t *)cause);
```
@andrewboie has already pointed the link to the documentation that explains how to proper validate parameters, take a look there. In essence you have to use a local copy  and then copy to user using `z_user_to_copy`.

arvinf replied:
Thank you @ceolin. This system call writes the result to `cause`. We don't use the user passed value. This is similar behaviour as `hwinfo_get_device_id` so I followed the same validation. What am I missing?

ceolin replied:
That one should be changed as well, take a look in https://docs.zephyrproject.org/latest/reference/usermode/syscalls.html#verification-memory-access-policies

arvinf replied:
That documentation states that local copying should be done if we read the parameter value, since the user can modify that between validation and use. For this call we're not. We're only copying data out. The example discourages use of `Z_SYSCALL_MEMORY_WRITE` if we're going to read the data. We are not.

Having said that, I don't mind following your suggestion and changing it to use the local copy. I just want to make sure I am understanding this correctly, because I am new to Zephyr development. Thanks again!

andrewboie replied:
Hi,
The documentation might be unclear. But we don't want to pass any user pointers to implementation functions except for the documented exception of large buffers (which this isn't.)
If `cause` is an output-only parameter, pass a stack variable address to the implementation function, and z_user_to_copy() it afterwards.

arvinf replied:
Thank you. I'll make the change.

henrikbrixandersen commented:
How come this should be part of the `hwinfo` driver API? Not all platforms provide reset cause and hardware identification through the same device...

alexanderwachter commented:
@henrikbrixandersen Because it is a hardware information. We decided to collect all hardware information in one API and therefor extend hwinfo if needed. As long as it only reads registers and does not need a device instance, I'm ok with putting it into hwinfo.

alexanderwachter commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -86,11 +86,139 @@ static void test_device_id_enotsup(void)
+ *   -# Reset cause value should change after calling clear reset cause.
+ * @}
+ */
+#ifdef CONFIG_HWINFO_HAS_DRIVER
```
Can you split `CONFIG_HWINFO_HAS_DRIVER` to `CONFIG_HWINFO_DEVICE_ID_HAS_DRIVER` and  `CONFIG_HWINFO_RESET_CAUSE_HAS_DRIVER`

FRASTM reviewed (commented):
Interesting feature to control the source of a reset. 
STM32 socs also have a information (a SBF bit in the PWR register) to check if the system resumed from Standby low power mode.

arvinf commented on drivers/hwinfo/hwinfo_stm32.c:
```diff
@@ -29,3 +29,72 @@ ssize_t z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+	if (LL_RCC_IsActiveFlag_PWRRST()) {
+		flags |= RESET_POWER;
+	}
+#endif
```
Correct. There will be some platform specific behaviour which the user will have to account for in her code. 

I considered setting both flags for this platform, but then you could a false trigger of brownout code path in the user's code when it was a POR. It can go both ways because the SoC flag itself is ambiguous. I decided to keep it simple.

arvinf commented on drivers/hwinfo/hwinfo_stm32.c:
```diff
@@ -29,3 +29,72 @@ ssize_t z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
+int z_impl_hwinfo_clear_reset_cause(void)
+{
+	__HAL_RCC_CLEAR_RESET_FLAGS();
+
```
I can make that change.

arvinf commented:
@alexanderwachter requested changes have been made.

alexanderwachter commented on include/drivers/hwinfo.h:
```diff
@@ -54,6 +70,26 @@ __syscall ssize_t hwinfo_get_device_id(u8_t *buffer, size_t length);
+ *
+ * @param      cause  OR'd flags TBD
+ *
+ * @return     Zero if successful otherwise (negative) error code.
```
```
* @retval  Zero if successful.
* @retval -ENOTSUP if there is no implementation for the particular device.
* @retval any negative value on driver specific errors.
```

alexanderwachter commented on include/drivers/hwinfo.h:
```diff
@@ -54,6 +70,26 @@ __syscall ssize_t hwinfo_get_device_id(u8_t *buffer, size_t length);
+/**
+ * @brief      Clears the reason for device reset.
+ *
+ * @return     Zero if successful otherwise (negative) error code.
```
```
* @retval  Zero if successful.
* @retval -ENOTSUP if there is no implementation for the particular device.
* @retval any negative value on driver specific errors.
```

alexanderwachter commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -84,13 +84,140 @@ static void test_device_id_enotsup(void)
+
+	ret = hwinfo_get_reset_cause(&cause);
+	/* If a platform does not support reset cause, pass the test. */
+	if (ret == -ENOTSUP) {
```
The driver should not return -ENOTSUP.
This is the test for `CONFIG_HWINFO_RESET_CAUSE_HAS_DRIVER`
Please make this an assert

arvinf replied:
Right. This was before we had a `CONFIG_` for each feature. I'll make the change.

alexanderwachter commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -84,13 +84,140 @@ static void test_device_id_enotsup(void)
+
+	ret = hwinfo_get_reset_cause(&cause_1);
+	/* If a platform does not support reset cause, pass the test. */
+	if (ret == -ENOTSUP) {
```
Same here

alexanderwachter reviewed (changes requested):
Some small things, Otherwise it looks good.

pabigot commented on include/drivers/hwinfo.h:
```diff
@@ -54,6 +70,30 @@ __syscall ssize_t hwinfo_get_device_id(u8_t *buffer, size_t length);
+/**
+ * @brief      Clear cause of device reset.
+ *
+ * @retval zero if successful.
```
This might not be the right API for clearing.  In my applications I need the ability to clear certain reasons, leaving the others in-place for the next reset.

arvinf replied:
That depends on the SoC. Some will clear the entire reason register as soon as it is read. Others let you clear specific bits.

pabigot reviewed (commented):
This is new API, and I don't personally believe it can be usefully generalized across SoCs, but for it to be acceptable it needs a wide variety of implementations to confirm that the behavior and reason management is consistent with what vendors provide.

https://github.com/zephyrproject-rtos/zephyr/pull/22132#issuecomment-577715449 has some background on my position.  I still have a requirement to fully control reset inspection and clearing in coordination with other things done on system startup; as long as nothing automatically invokes these functions (including vendor startup code) that requirement can be satisfied.

arvinf commented on include/drivers/hwinfo.h:
```diff
@@ -30,6 +30,22 @@
 extern "C" {
 #endif
 
+enum reset_cause {
```
That's a valid point. More information is useful. There will be some platform specific behaviour so the documentation should be as generic as possible.

arvinf commented:
Addressing your comment in #22132

> What is the likelihood that there will be a cross-platform real application that would benefit from having a generic API to inspect and clear reset status?

I had that very specific need which is why I added the API. It needs to work across 3 different platforms. This PR supports one. I'll be submitting the other two shortly.

Capturing diverse platforms in an API is always going to be difficult as you observed, but it doesn't mean it shouldn't be done. Sensor API and even PWM API are examples of that. The API as is should satisfy most use cases, out of the box. Reset causes, broadly, are:
- Software
- POR/BOR
- Watchdog

Most platforms support those. There will be some other cases supported by platforms like CPU lock up, etc. which we can support in the enum and the user can consult their SoC documentation for more info on those flags.

alexanderwachter commented:
@nordic-krch Can you comment on what reset causes would be usefull for the NRF series?

pabigot commented:
Nordic reset reasons include:
* RESETPIN - Reset from pin-reset detected
* DOG - Reset from watchdog detected
* SREQ - Reset from soft reset detected
* LOCKUP - Reset from CPU lock-up detected
* OFF - Reset due to wake up from System OFF mode when wakeup is triggered from DETECT signal from GPIO
* LPCOMP - Reset due to wake up from System OFF mode when wakeup is triggered from ANADETECT signal from LPCOMP
* DIF - Reset due to wake up from System OFF mode when wakeup is triggered from entering into debug interface mode
* NFC - Reset due to wake up from System OFF mode by NFC field detect
* VBUS - Reset due to wake up from System OFF mode by VBUS rising
into valid range

alexanderwachter commented:
@arvinf Do you think it makes sense to generalize it a little bit more? Or even use two flags? One generalized and one SoC dependent?
With generalizing, I mean only one watchdog, only one peripheral, only one pin, only one bor, ...

arvinf commented:
I'd like to use the general flags as much as possible. Let me think on this a little more -- and do some research on different SoC reset causes out there (or at least ARM ones) -- for the chips Zephyr already supports to see how large the flag space is to begin with. It may be that we can support all the flags out there in the existing `enum` with room to spare.

alexanderwachter commented:
@arvinf The problem I see when having multiple or specialized flags for things like a watchdog is that the API becomes SoC depending. The user then needs to know which SoC he is compiling for to check for the right flags. If that is going to be the case, you could use the vendor HAL directly.

arvinf commented:
You are right that it becomes SoC dependant to some extent and they *could* use the HAL directly. However as I see it we have to provide the general flags, and might as well provide the remaining flags instead of asking the user to call their HAL. 

We also run the risk of creating problems for the user if she wants to retrieve the SoC specific flags since in some SoCs the reason register is cleared by reading it and can't be restored.

Your suggestion of having a general flag and an extended flag is not a bad idea. I want to first figure out the available flags and their overlap so we have a better idea of what is general and what is not.

pabigot commented:
> However as I see it we have to provide the general flags, and might as well provide the remaining flags instead of asking the user to call their HAL.

No, we do not have to provide a Zephyr API for this functionality at all.

If we cannot come up with a generic (SOC-independent) API that meets the requirements of enough applications to be useful then we should direct people to use their own abstractions if they require cross-platform support for detecting reset causes.

arvinf commented:
These sources are available on virtually all SoCs:
- Reset Pin
- Software Reset
- Watchdog
- POR
- BOR
- JTAG/Debug port

These are available on a good number of SoCs (low-power, higher numbered Cortex-M) but not all:
- CPU Lockup
- Lowpower Wake
- Lowpower Error
- Security Firewall

The remaining flags are very SoC specific.

There are common, reasonable use cases for the above in a variety of cross platform applications.

The main API can return the common flags. If a platform supports multiple BOR or Watchdog sources for example they would all map to the same BOR flag in the API. The user can retrieve extended flags through the SoC HAL. Alternatively  we can also provide a call to retrieve the raw register values which can then be parsed by the user on an SoC specific basis. 

I prefer the latter because if we have a platform that clears reset source register on read then the user can't call zephyr API and the HAL. We can however cache the raw values on the first read and return them to the user on subsequent calls.

henrikbrixandersen commented:
> If we cannot come up with a generic (SOC-independent) API that meets the requirements of enough applications to be useful then we should direct people to use their own abstractions if they require cross-platform support for detecting reset causes.

I tend to agree with @pabigot on this. I fear we will end up with a too generic API for this very SoC specific functionality - an API, that will not be sufficient for real-life applications, and thus won’t see widespread use.

arvinf commented:
I've outlined a number of flags that are applicable to virtually all SoCs out there. They can be supported by a generic API.

alexanderwachter commented:
I think we should go with a generic API as described with the flags above. But we should at least provide drivers for the most important SoCs like NXP, Atmel, Nordic and STM, to check if we missed something.

arvinf commented:
I have drivers for 3 platforms. I am going to update this PR in the next few days with those implementations. I'll update the enum with the generic flags, as above.

nordic-krch commented:
At least on nordic SoCs reading reset reason does not clear it and register is accumulative. User must explicitly write to clear given reset reason. If there is a bootloader that reads the reset reason it should not clear it (in typical case) but application should probably read and clear. Also within application there might be multiple modules which would like to read reset reason so it should be stored somewhere as well.

arvinf commented:
The proposed API has a specific call `hwinfo_clear_reset_cause ` to clear the cause after reading. For SoCs that are accumulative this works out of the box. For SoCs that clear the register on read, the value must be cached internally on the first read and the returned on subsequent calls; `hwinfo_clear_reset_cause` would clear the locally cached copy.

nordic-krch commented:
then how do we foresee bootloader case? There might be cases when bootloader needs to read the reset reason without clearing it (e.g. to skip integrity check on wake-up from deep sleep).

arvinf commented:
There needs to be another mechanism to pass values between BL and the app. That's a more general message passing problem between different apps in a multiapp image. Something like passing command line to the kernel. That feature is very useful but beyond the scope of this PR.

carlescufi commented:
I will bring this up in the next API meeting. Feel free to join @arvinf . I will send an agenda beforehand.

nashif commented:
this PR seems to still be using the old types

arvinf commented:
I was holding off until we're closer to merge window, in case there are other changes needed.

pabigot reviewed (dismissed):
I'm registering the request changes simply because there are at least three people who have commented with reservations about the usefulness of this API, but haven't explicitly rejected it, and I don't recall that it ever did get discussed in an API meeting.  I think it warrants that level of attention. 
  @carlescufi @henrikbrixandersen @nordic-krch 

I still feel this is something for which there's negative value in trying to come up with a cross-vendor solution abstraction, but if after we discuss it the general consensus is this is useful, and I can confirm that it's easy to completely disable it in both application and bootloader, I won't block it.

carlescufi commented:
@pabigot I see no hard objections to adding this API, although I would ask @anangl, @MaureenHelm and @henrikbrixandersen to weigh in (again) if possible. From my side, while I recognize the limitations in trying to generalize something like a reset reason, I still think this has the potential to be useful to a big number of applications.

I will add this to next week's API meeting.

erwango commented:
@carlescufi we're also in favor of this API. Particularly it is useful in PM cases to identify wake up sources in various sleep states.

galak commented on include/drivers/hwinfo.h:
```diff
@@ -30,6 +30,18 @@
+	RESET_SECURITY				= BIT(6),
+	RESET_LOW_POWER_WAKE			= BIT(7),
+	RESET_CPU_LOCKUP			= BIT(8)
+};
```
This needs far more docs on what these are suppose to mean.

galak replied:
Probably adding a RESET_SOC (ie SoC specific reason).

arvinf replied:
Agreed. I am trying to gather all feedback before doing more work. There was/is some uncertainty about the overall approach.

arvinf replied:
Would RESET_OTHER work better than RESET_SOC for the causes that are not supported by this list?

carlescufi commented on include/drivers/hwinfo.h:
```diff
@@ -30,6 +30,18 @@
 extern "C" {
 #endif
 
+enum reset_cause {
```
This should be a set of `#define` statements instead of an `enum`

mnkp replied:
This old comment hasn't been addressed yet. We cannot use enum here, `cause` parameter is of type `uint32_t*`.

carlescufi commented on include/drivers/hwinfo.h:
```diff
@@ -54,6 +66,30 @@ __syscall ssize_t hwinfo_get_device_id(uint8_t *buffer, size_t length);
+ * @brief      Retrieve cause of device reset.
+ *
+ * @param      cause  OR'd flags TBD
+ *
```
Add a warning that states that on some MCUs the mere fact or reading the flags will also clear them, which is a problematic side-effect.

carlescufi replied:
Alternatively, implementations that suffer from this limitation could cache the value of the reset reason so that the value can be read using this call multiple times. This wouldn't work with a bootloader, but it would work if this API is called multiple times inside the main image.

carlescufi commented on include/drivers/hwinfo.h:
```diff
@@ -54,6 +66,30 @@ __syscall ssize_t hwinfo_get_device_id(uint8_t *buffer, size_t length);
+ * @retval any negative value on driver specific errors.
+ */
+__syscall int hwinfo_clear_reset_cause(void);
+
```
@galak proposes an additional call that returns the bitmask of supported reasons in a particular MCU.

stephanosio commented on include/drivers/hwinfo.h:
```diff
@@ -30,6 +30,18 @@
 #endif
 
+enum reset_cause {
+	RESET_PIN				= BIT(0),
+	RESET_SOFTWARE				= BIT(1),
+	RESET_BROWNOUT				= BIT(2),
+	RESET_POWER				= BIT(3),
+	RESET_WATCHDOG				= BIT(4),
+	RESET_DEBUG				= BIT(5),
+	RESET_SECURITY				= BIT(6),
+	RESET_LOW_POWER_WAKE			= BIT(7),
+	RESET_CPU_LOCKUP			= BIT(8)
```
Any reason to have both `RESET_WATCHDOG` and `RESET_CPU_LOCKUP`? In general, "CPU lockup" would almost always mean the same thing as "reset by watchdog."

Also there should be something like `RESET_ECC_FAIL` since this is a quite common feature found in the (automotive and industrial) safety MCUs.

arvinf replied:
I came up with this list by looking through the docs for most/all the SoCs that are supported by Zephyr. CPU Lockup is different than Watchdog. It is asserted by ARM cores when there is an error while executing NMI or HardFault handlers. 

I will add ECC/Parity flags if you can point me to an example SoC.

stephanosio replied:
> I came up with this list by looking through the docs for most/all the SoCs that are supported by Zephyr and selecting the best common denominator. CPU Lockup is different than Watchdog. It is asserted by ARM cores when there is an error while executing NMI or HardFault handlers.

So, I suppose a more generic and technically correct name would be "CPU fault" then. AFAICS, the term "CPU lockup" in this context is only used by the ARM docs and a handful of SoCs.

> I will add ECC/Parity flags if you can point me to an example SoC.

* Atmel SAM D/E 5x series (e.g. ATSAME54)
* Atmel SAM E/V 7x series (e.g. ATSAME70)
* TI Hercules family (e.g. TMS570, RM)

arvinf replied:
I kept the name same as the docs, but making it generic is not a bad idea.

> Atmel SAM D/E 5x series (e.g. ATSAME54)
Atmel SAM E/V 7x series (e.g. ATSAME70)
TI Hercules family (e.g. TMS570, RM)

👍

alexanderwachter commented:
@arvinf, can you write some words about the design, design decisions, and warning about resetting while reading the registers of this new API? The doc is in [zephyr/doc/reference/peripherals/hwinfo.rst](https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/reference/peripherals/hwinfo.rst). Some example code would also be helpful. The doc is published [here](https://docs.zephyrproject.org/latest/reference/peripherals/hwinfo.html#hwinfo-api)

stephanosio commented on include/drivers/hwinfo.h:
```diff
@@ -30,6 +30,18 @@
+enum reset_cause {
+	RESET_PIN				= BIT(0),
+	RESET_SOFTWARE				= BIT(1),
+	RESET_BROWNOUT				= BIT(2),
+	RESET_POWER				= BIT(3),
```
Should this be `RESET_POWER_ON`, especially noting that there are other "power"-related events such as `RESET_BROWNOUT` and `RESET_LOW_POWER_WAKE`.

arvinf replied:
Probably better to go with `RESET_POR`

alexanderwachter commented:
@arvinf, do you want to continue working on this PR, or should we close it?

arvinf commented:
Please keep it open. Just haven't found the time to write up the docs.

cfriedt commented on tests/drivers/hwinfo/api/testcase.yaml:
```diff
@@ -1,3 +1,7 @@
 tests:
   drivers.device_id:
     tags: driver
+  drivers.get_reset_cause:
```
Need a userspace test config too.
See `tests/kernel/mutex/sys_mutex/testcase.yaml`

cfriedt commented on drivers/hwinfo/hwinfo_weak_impl.c:
```diff
@@ -10,3 +10,13 @@ ssize_t __weak z_impl_hwinfo_get_device_id(u8_t *buffer, size_t length)
 {
 	return -ENOTSUP;
 }
+
```
I think even the weak implementation can provide some greater functionality, although it would also make sense if that were done in a different PR.

For example, if a device does not support a reset-reason register, then a few `uint32_t`s or something along those lines could be queried for flags & magic values. Just so that it isn't zero'ed with the rest of BSS it would need to go into a separate linker section for persistent data.

Clearly, this may have noise when a device is powered on, hence the `MAGIC` values in [reboot(2)](https://man7.org/linux/man-pages/man2/reboot.2.html). I don't know if Zephyr has a "reboot" system call already.. does it?

As long as the CRT doesn't wipe values, they can can persist between reboots, even when using DRAM if the system doesn't lose power (which is how Android handles reboot reasons, IIRC).

Someone already mentioned that the reset reason could be passed from BL to Zephyr with some notion of "command line arguments" or something similar, but that's also out of scope here.

cfriedt replied:
I would say something like that even has value when the arch does have a reset-reason register, because at least everything would be kind of forced to fit into a vendor-neutral model.
My 2c though.

Definitely an extremely useful feature!

carlescufi commented:
> Please keep it open. Just haven't found the time to write up the docs.

@arvinf do you plan to resume work on this PR? The API meeting consensus seems to be that this would be a welcome addition if we can resolve the outstanding issues.

arvinf commented:
I've improved the documentation in the driver header. I've added `hwinfo_get_supported_reset_cause` based on the feedback here and rebased onto the latest master.

cfriedt commented:
Looks like there may be some STM32 platform-specific headers missing. Otherwise, it seems to be getting there. There are a few comments that need to be addressed still.

mnkp commented on drivers/hwinfo/Kconfig:
```diff
@@ -10,7 +10,10 @@ menuconfig HWINFO
+config HWINFO_DEVICE_ID_HAS_DRIVER
+	bool
+
+config HWINFO_RESET_CAUSE_HAS_DRIVER
```
We have weak implementations so there is no good reason to introduce yet another Kconfig option. It complicates the code but brings no benefit.

arvinf replied:
This was requested [earlier](https://github.com/zephyrproject-rtos/zephyr/pull/24884#discussion_r419265245).

alexanderwachter replied:
@mnkp the difference is that the `CONFIG_HWINFO_HAS_DRIVER` can be checked at compile time.

mnkp replied:
Sorry, didn't notice the earlier request. This approach however is not clean. We have and should stay with one driver called hwinfo. With this setup it's not really clear what reset cause really is. It looks like we have two separate drivers now, one called "device id" and second called "reset cause". Even if they are packed in the same implementation file. Looks messy.

We could implement it as an extension of features of hwinfo driver, i.e. a Kconfig option would enable extra features in the hwinfo API. Though I still see no good reason. This extra option doesn't allow to write portable applications since the application (assuming the same code would run on different SoC families) would still need to call `hwinfo_get_reset_cause` function at runtime to verify that the reset cause it requires is supported by the hardware.

alexanderwachter replied:
@mnkp when we introduced the hwinfo API, we agreed that it should exactly look how it is now. We had discussions in the API meeting and agreed on this approach.
However, the `CONFIG_*_HAS_DRIVER` is a nice to have and IMO we should keep it. Maybe someone wants to have static asserts in his code. The thing is that hwinfo has no device-tree and therefor Kconfig is the only way to her compile-time information about the availability.

mnkp replied:
I may be missing some arguments that were raised earlier. Could we maybe have a short discussion during next API meeting? And if not let's leave it as is.

henrikbrixandersen replied:
The concerns raised by @mnkp here matches the concerns I raised in this PR (https://github.com/zephyrproject-rtos/zephyr/pull/24884#issuecomment-623309496) a year ago.

arvinf replied:
Whatever your decide after your internal discussion please let me know and I'll make the changes appropriately.

alexanderwachter commented:
@carlescufi can you put it to the agenda?

alexanderwachter commented:
@mnkp seems that we spend too much time in pinmux discussion in the API meeting :-/
Still I want to get this in 2.6.
Either we discuss here, accept and change later if needed or use some other meeting.
cc @carlescufi

galak commented:
> @mnkp seems that we spend too much time in pinmux discussion in the API meeting :-/
> Still I want to get this in 2.6.
> Either we discuss here, accept and change later if needed, or use some other meeting.
> cc @carlescufi

We can discuss in dev-review meeting on Thur.  @mnkp, @alexanderwachter please try and join.

carlescufi commented:
> @carlescufi can you put it to the agenda?

I am sorry I missed this, I agree with @galak, let's discuss it today in the dev-review.

galak commented:
Get the Kconfig compliance issue get fixed.

carlescufi commented:
> @carlescufi can you put it to the agenda?

@alexanderwachter sorry, I saw this too late for this week. Next time please ping me directly on Slack, I don't see GitHub notifications timely enough.

arvinf commented:
@carlescufi - done.

henrikbrixandersen commented on drivers/hwinfo/hwinfo_nrf.c:
```diff
@@ -9,6 +9,9 @@
 #include <string.h>
 #include <hal/nrf_ficr.h>
 #include <sys/byteorder.h>
+#ifndef CONFIG_BOARD_QEMU_CORTEX_M0
```
Is this only for QEMU? Or should it check on a given nRF family instead?

arvinf replied:
QEMU only.

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -35,7 +34,10 @@ static void test_device_id_get(void)
 	length_read_1 = hwinfo_get_device_id(buffer_1, 1);
-	zassert_not_equal(length_read_1, -ENOTSUP, "Not supported by hardware");
+	if (length_read_1 == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -64,33 +66,135 @@ static void test_device_id_get(void)
+
+	ret = hwinfo_get_reset_cause(&cause);
+	if (ret == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -64,33 +66,135 @@ static void test_device_id_get(void)
+
+	ret = hwinfo_get_reset_cause(&cause_1);
+	if (ret == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -64,33 +66,135 @@ static void test_device_id_get(void)
+
+	ret = hwinfo_clear_reset_cause();
+	if (ret == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -64,33 +66,135 @@ static void test_device_id_get(void)
+
+	ret = hwinfo_get_reset_cause(&cause_2);
+	if (ret == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

henrikbrixandersen commented on tests/drivers/hwinfo/api/src/main.c:
```diff
@@ -64,33 +66,135 @@ static void test_device_id_get(void)
+
+	ret = hwinfo_get_supported_reset_cause(&supported);
+	if (ret == -ENOSYS) {
+		return;
```
```suggestion
		ztest_test_skip();
		return;
```

alexanderwachter commented on drivers/hwinfo/hwinfo_handlers.c:
```diff
@@ -14,3 +14,34 @@ ssize_t z_vrfy_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+}
+#include <syscalls/hwinfo_clear_reset_cause_mrsh.c>
+
+ssize_t z_vrfy_hwinfo_get_supported_reset_cause(uint32_t *supported)
```
```suggestion
int z_vrfy_hwinfo_get_supported_reset_cause(uint32_t *supported)
```

alexanderwachter commented on drivers/hwinfo/hwinfo_handlers.c:
```diff
@@ -14,3 +14,34 @@ ssize_t z_vrfy_hwinfo_get_device_id(uint8_t *buffer, size_t length)
 }
 #include <syscalls/hwinfo_get_device_id_mrsh.c>
+
+ssize_t z_vrfy_hwinfo_get_reset_cause(uint32_t *cause)
```
```suggestion
int z_vrfy_hwinfo_get_reset_cause(uint32_t *cause)
```

alexanderwachter commented on drivers/hwinfo/hwinfo_handlers.c:
```diff
@@ -14,3 +14,34 @@ ssize_t z_vrfy_hwinfo_get_device_id(uint8_t *buffer, size_t length)
+#include <syscalls/hwinfo_get_reset_cause_mrsh.c>
+
+
+ssize_t z_vrfy_hwinfo_clear_reset_cause(void)
```
```suggestion
int z_vrfy_hwinfo_clear_reset_cause(void)
```

alexanderwachter reviewed (changes requested):
Almost there! Just some minor changes. (Copy & paste error, I guess)

arvinf commented:
@carlescufi  - Could you please let me know if there is anything else I need to do for this PR to be merged?

alexanderwachter commented:
@arvinf we are in code-freeze for 2.6 atm. We will merge after the release

cfriedt reviewed (approved):
Nice! I'm happy this got through :-)
